main/src/ca/uwaterloo/flix/language/ast/Type.scala

```diff
@@ -313,9 +307,24 @@ object Type {
   /////////////////////////////////////////////////////////////////////////////
 
   /**
-    * A type variable expression.
+    * The union of type variables.
     */
-  case class Var(id: Int, kind: Kind, rigidity: Rigidity = Rigidity.Flexible, text: Option[String] = None) extends Type with Ordered[Type.Var] {
+  sealed trait MaybeKindedVar extends Type {
+    def id: Int
+    def text: Option[String]
+  }
+
+  /**
+    * The union of non-Apply types.
+    * Used to restrict the range of return values of [[Type.baseType]].
+    */
+  sealed trait BaseType extends Type
```

Comment:
I'd like to know the reasoning for having this base type.

Reply:
If I call `tpe.baseType` I get a `BaseType` result. Then I don't need a `case _: Type.Apply => throw InternalCompilerException("unexpected apply type")` in the pattern match.